Add blog post: When Ralph Wiggum Meets Pulumi (#17049)
pulumi/docs: content/blog/when-ralph-wiggum-meets-pulumi/index.md

---

> title: "When Ralph Wiggum Meets Pulumi"

Reviewer: ```suggestion
title: "How Ralph Wiggum Built a Serverless SaaS with Pulumi"
```

allow_long_title: true
date: 2026-01-13T09:00:00-08:00
draft: false

> meta_desc: "How the Ralph Wiggum autonomous loop technique built a complete serverless URL shortener SaaS on AWS using Pulumi without supervision."

Reviewer: ```suggestion
meta_desc: "See how an AI agent using the Ralph Wiggum loop built a full AWS serverless SaaS with Pulumi—unsupervised. Frontend, backend, tests, and CloudFront included."
```

done
```

That is it. That is the whole thing. You write your instructions in a `PROMPT.md` file, and the loop keeps feeding it to Claude Code over and over.

But here is why it works: each time Claude starts up, it looks at the project. It sees the files that exist, the code that is already written, the git history. It is not starting from zero. It picks up where it left off, sees what still needs to be done, and keeps going.

The official Claude Code plugin formalizes this with some important additions:

- **Completion promises**: You tell Claude to output a specific string like `<promise>COMPLETE</promise>` when it has genuinely finished
- **Max iterations**: A safety limit so your API bill does not reach infinity
- **Stop hooks**: The mechanism that intercepts exit attempts and re-injects the prompt

To install it:

```bash
/plugin install ralph-wiggum@ghuntley
```

For end-to-end testing of deployed infrastructure, you will also want the Playwright MCP server. This lets Claude interact with your deployed application in a real browser:

```bash
claude mcp add playwright npx @playwright/mcp@latest
```

## Why Pulumi is perfect for Ralph

> Infrastructure as code has something most application code does not: objective success criteria. Your Lambda either deploys or it does not. Your DynamoDB table either exists or it does not. Tests either pass or they fail.

Reviewer: ```suggestion
[Infrastructure as code](/what-is/infrastructure-as-code/) has something most application code does not: objective success criteria. Your Lambda either deploys or it does not. Your DynamoDB table either exists or it does not. Tests either pass or they fail.
```
